xarray/ufuncs.py

```diff
@@ -0,0 +1,125 @@
+"""xarray specific universal functions."""
+
+import textwrap
+import warnings
+
+import numpy as np
+
+import xarray as xr
+from xarray.core.groupby import GroupBy
+from xarray.namedarray.pycompat import array_type
+
+
+def _walk_array_namespaces(obj, namespaces):
+    if isinstance(obj, xr.DataTree):
+        for node in obj.subtree:
+            _walk_array_namespaces(node.dataset, namespaces)
+    elif isinstance(obj, xr.Dataset):
+        for name in obj.data_vars:
+            _walk_array_namespaces(obj[name], namespaces)
+    elif isinstance(obj, GroupBy):
+        _walk_array_namespaces(next(iter(obj))[1], namespaces)
+    elif isinstance(obj, xr.DataArray | xr.Variable):
+        _walk_array_namespaces(obj.data, namespaces)
+    elif isinstance(obj, array_type("dask")):
+        _walk_array_namespaces(obj._meta, namespaces)
+    else:
+        namespace = getattr(obj, "__array_namespace__", None)
+        if namespace is not None:
+            namespaces.add(namespace())
+
+    return namespaces
+
+
+class _UFuncDispatcher:
+    """Wrapper for dispatching ufuncs."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, *args, **kwargs):
+        xps = set()
+        for arg in args:
+            _walk_array_namespaces(arg, xps)
+
+        xps.discard(np)
+        if len(xps) > 1:
+            names = [module.__name__ for module in xps]
+            raise ValueError(
+                f"Mixed array types {names} are not supported by xarray.ufuncs"
+            )
+
+        xp = next(iter(xps)) if len(xps) else np
+        func = getattr(xp, self._name, None)
+
+        if func is None:
+            warnings.warn(
+                f"Function {self._name} not found in {xp.__name__}, falling back to numpy",
+                stacklevel=2,
+            )
+            func = getattr(np, self._name)
```

Comment:
I would lean towards skipping this fall-back, unless there are particularly motivating cases.

Reply:
Motivation here would be duck arrays that implement `__array_ufunc__` and don't implement the full suite of numpy ufuncs. I ran into this with sparse. Not sure the full delta list, but I see they don't have sin/cos for example. In this case, `np.cos(x_sparse)` works but `xp.cos(x_sparse)` fails, which is a little weird. Not the most elegant solution though, I agree.

Reply:
I think it's intentional than `xp.cos(x_sparse)` fails, because `cos(0) != 0`, so the result is no longer sparse.

Reply:
Actually I was wrong, there is a `sparse.cos` and a bunch of others, although they don't appear in the API docs. It seems sparse's general approach to these is compute elementwise on the valid data, and then modify the fill_value as required for the empty regions.

There are still 40-some functions that fail without this fallback, although generally more niche. With the fallback, all work and output a sparse array (no auto densification):

```
absolute, arccos, arccosh, arcsin, arcsinh, arctan, arctan2, arctanh, bitwise_count, cbrt, conj, conjugate, copysign, deg2rad, degrees, exp2, expm1x, fabs, float_power, fmax, fmin, fmod, gcd, heaviside, hypot, invert, isreal, lcm, ldexp, left_shift, logaddexp2, maximum, minimum, mod, nextafter, power, rad2deg, radians, reciprocal, right_shift, rint, signbit, spacing, true_divide
```